server/routes/companies.js

```diff
@@ -116,25 +179,459 @@ router.get('/filter', function(req, res, next) {
+    .leftOuterJoin('company_position', 'company_major.company_id', 'company_position.company_id')
+    .where({position_id: req.body.pid, major_id: req.query.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (mid) {
-    db.query('SELECT company_id FROM company_major WHERE major_id = $1;', [mid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_major').select('company_id').where({major_id: req.body.mid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else if (pid) {
-    db.query('SELECT company_id FROM company_position WHERE position_id = $1;', [pid], (err, result) => {
-      if (err) return next(err);
-      res.send(result.rows);
-    });
+    knex('company_position').select('company_id').where({position_id: req.body.pid})
+      .then(result => {
+        if (result.length) {
+          res.json(result);
+        } else {
+          util.throwError(404, 'No companies found');
+        }
+      })
+      .catch(err => { return next(err) });
   } else { 
-    return res.status(400).send({
-      message: '\'mid\' and \'pid\' are undefined'
-    });
+    util.throwError(400, '\'mid\' and \'pid\' are undefined');
+  }
+});
+
+// Add a single company or related data
+router.post('/', function(req, res, next) {
+  values = {
+    name: req.body.name,
+    website: req.body.website,
+    logo: req.body.logo,
+    citizenship_requirement: req.body.citizenship_requirement,
+    description: req.body.description
+  }
+
+  let ranks = req.body.rank;
+  let user_ids = req.body.user_id;
+  let position_ids = req.body.position_id;
+  let major_ids = req.body.major_id;
+  let contact_ids = req.body.contact_id;
+  let event_ids = req.body.event_id;
+
+  let rankValues = [];
+  if (ranks && user_ids) {
+    if(Array.isArray(ranks) && Array.isArray(user_ids)){
+      let length = ranks.length;
+      if (ranks.length != user_ids.length)
+        util.throwError(400, "Both user_ids and ranks must be the same size.");
+      length = user_ids.length; 
+      for (let n=0; n < length; n++) {
+        rankValues.push({
+          user_id: user_ids[n],
+          rank: ranks[n]
+
+        });
+       // util.throwError(400, ranks)
+      }
+    } else if(Array.isArray(ranks) || Array.isArray(user_ids)) {
+       util.throwError(400, "Both user_ids and ranks must be the same size.");
+    }else{
+      rankValues.push({
+          user_id: user_ids,
+          rank: ranks
+        });
+    }
+
   }
+
+  let companyPositions = [];
+  if (position_ids) {
+    if(Array.isArray(position_ids)){
```

Comment:
To make these loops more concise use .forEach()

![image](https://user-images.githubusercontent.com/17267100/57701776-89cdd380-7611-11e9-981c-d13926d6b829.png)

Reply:
Also keep the companyMajors.push({ major_id: major_ids[n] }) on one line to help conciseness